cmd/clusterctl/cmd/init.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+)
+
+type initOptions struct {
+	kubeconfig              string
+	coreProvider            string
+	bootstrapProviders      []string
+	infrastructureProviders []string
+	targetNamespace         string
+	watchingNamespace       string
+	force                   bool
+}
+
+var io = &initOptions{}
+
+var initCmd = &cobra.Command{
+	Use:   "init",
+	Short: "Initialize a management cluster for Cluster API",
+	Long: LongDesc(`
+		Initialize a management cluster for Cluster API by installing cluster API core components,
+		the kubeadm bootstrap provider, and the selected bootstrap and infrastructure providers.
+
+		The management cluster should be an existing Kubernetes cluster, and the identity used
+		for accessing the cluster must have enough privileges for installing cluster API providers.
+
+		Use 'clusterctl config providers' to get the list of available providers; if necessary, edit
+		the .clusterctl file to add new provider configurations or to customize existing ones.
+		
+		Some providers require environment variable to be set before running clusterctl init; please
+		refer to the provider documentation or use 'clusterctl config provider [name]' to get the
+		list of required variables.
+
+		See https://cluster-api.sigs.k8s.io/ for more details.`),
+
+	Example: Examples(`
+		# Initialize a management cluster, the cluster currently pointed by kubectl, by installing the
+		# AWS infrastructure provider. Please note that this command, when executed on an empty management cluster, 
+ 		# automatically triggers the installation of the cluster-api core provider.
+		clusterctl init --infrastructure=aws
+
+		# Initialize a management cluster with a specific version of the AWS infrastructure provider.
+		clusterctl init --infrastructure=aws:v0.4.1
```

Comment:
I think `@` is more synonymous with versioning, thoughts on:
```
clusterctl init --infrastructure=aws@v0.4.1
```
?

Reply:
Also :version from docker is quite common nowadays... But I don't have strong opinions, ready to change this if there are more request in this direction

Reply:
go favors the `@` symbol, docker the `:`.

We could consider an `--infrastructure-version` flag as an alternative. When the `--infrastructure-version` flag is omitted, `latest` release is used. (`latest` != `master`).

No preference on `@`, `:` or any other somewhat common separator

Reply:
Personally I prefer one of `:` or `@` vs adding a flag for the following reasons:
1. The flag looks weird when installing more than one provider. e.g . `--infrastructure=aws,vspere` and then  `--infrastructure-version=version_aws,version_vsphere`
2. If we add `--infrastructure-version `, then we should also `--infrastructure-version` and `--core-version `

Reply:
I propose we proceed - this is just a scaffolding. We will have room to iterate.

Reply:
Let's start with `:`, then eventually it will be easier to switch to `@`